[TensorExpr] Extend arithmetic simplifier to work with multi variable expressions

nickgg opened the pull request:
A new version of the IR simplifier used by the jit/tensorexpr fuser. This is capable of simplifying expressions containing (shock) multiple variables, eg:

```(m * (1 * n_1) + (n  + 1)) - (m *  (1 * n_1) + n) => 1```

Similar to the previous IR Simplifier it uses a two stage approach:
1. Traverse the tree combining subtree's of commutable operations in to a flat structure. In this implementation we have two intermediate Exprs: Term (expressing products of sub expressions) and Polynomial (expressing sums of sub expressions).
2. Traverse the tree expanding Term's and Polynomials into their component operators.

Using the example above we execute with a process like this to simplify:
```
   (m * (1 * n_1) + (n  + 1)) - (m *  (1 * n_1) + n)
# Using PolynomialTransformer:
=> Sub(Add(Mul(m, Mul(1, n_1)), Add(n, 1)), Add(Mul(m, Mul(1, n_1)), n))
=> Sub(Polynomial(Term(m, n_1), n, 1), Polynomial(Term(m, n_1), n))
=> Polynomial(Term(m, n_1), Term(-1, m, n_1), n, -n, 1)
=> Polynomial(1)
# Using TermExpander
=> 1
```

The IRSimplifier supports arithmetic simplifications of operators Add, Sub and Mul and constant folding of all binary Exprs and Intrinsics, but does not attempt expansion of multiplication of Polynomials to the canonical form since that generally leads to less efficient representations. It will do scalar factorization if it results in removal of operators, and will merge chains of multilane primitives (such as Broadcast and Ramp) down into a single operator. The ir_simplifier unit tests are a short tour of its capabilities.

The existing simplifier has a bug where it will sometimes reorder operations on floating point types which are not associative. This causes (at least) the pyhpc equation_of_state benchmark to produce incorrect results. I have fixed that issue in this version and verified that that benchmark produces the same results with and without the simplifier.

Tests: all cpp & py tensorexpr tests, and pyphc benchmark:
```
benchmarks.equation_of_state
============================
Running on CPU

size          backend     calls     mean      stdev     min       25%       median    75%       max   Δ
------------------------------------------------------------------------------------------------------------------
   4,194,304  pytorch           10     0.246     0.002     0.243     0.245     0.246     0.248     0.250     1.000
```

nickgg committed:
[TensorExpr] Multi variable arithmetic simplifier

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_simplifier.h:
```diff
@@ -9,17 +11,260 @@ namespace torch {
+      const Expr* s,
+      std::unordered_map<SimplifierHashType, const Expr*> varmap)
+      : ExprNodeBase(promoteTypesMap(s, varmap)), scalar_(s), hasher_(hasher) {
+    for (auto& p : varmap) {
```
We're iterating over an unordered map here, which results in random order of components in the term. Would it be possible to fix the order somehow? Even if the order doesn't matter from computational standpoint, it would be annoying to debug this piece as the dumps could differ for no reason.

nickgg replied:
The components are sorted below, so they end up in a consistent order.

nickgg replied:
Or are you concerned about debugging the call to addComponent itself?

ZolotukhinM replied:
> The components are sorted below, so they end up in a consistent order.
It's less of a problem then.

> Or are you concerned about debugging the call to addComponent itself?
I think it's fine to keep it as-is now then.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_simplifier.h:
```diff
@@ -9,17 +11,260 @@ namespace torch {
+  }
+
+  // Sort by hash.
+  void sort() {
```
Might be worth adding a comment why sorting by hash is a good idea (I guess it's because we only care about putting identical terms close to each other, but the order relative to other terms doesn't matter). Initially when I saw 'sort by hash', it alarmed me.

nickgg replied:
I'll add a comment, but the reason is just that the hasher is order dependent but we want to combine terms which are equal but reordered (e.g. x*y == y*x).

ZolotukhinM replied:
Yeah, I understand the motivation, it was just weird to see 'sort by hash'. In your example we really don't care how we sort `xy`: whether it ends up as `xy` or `yx` - and that was the weird part when I first saw the code (it's not what I expect from sorting of something). It's more of normalization than of sorting, maybe.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_simplifier.h:
```diff
@@ -9,17 +11,260 @@ namespace torch {
+};
+
+// Polynomial represents a grouping of Exprs by addition.
+// E.g. sum(*variables, scalar).
```
What is scalar here? Can the component be something else (neither Var nor FloatImm/IntImm)?

nickgg replied:
Scalar term must be Immediate, variable terms can be anything (Var, compound Expr of any ops, Polynomial or Term).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_simplifier.h:
```diff
@@ -9,17 +11,260 @@ namespace torch {
+      const Expr* s,
+      std::unordered_map<SimplifierHashType, const Term*> varmap)
+      : ExprNodeBase(promoteTypesMap(s, varmap)), scalar_(s), hasher_(hasher) {
+    for (auto& p : varmap) {
```
Similar question about iterating over unordered_map.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_simplifier.h:
```diff
@@ -9,17 +11,260 @@ namespace torch {
+// Uses the evaluator to fold an Expression with constant terms.
+// E.g. evaluateOp(Add(3, 4)) => 7.
+// Expr v must not have any unbound Vars.
 static Expr* evaluateOp(const Expr* v) {
```
Nit: this probably belongs to `eval.h`

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_simplifier.h:
```diff
@@ -63,304 +309,732 @@ static const Expr* newBinaryOpOfType(
-
-  return new Add(mul, B);
+// Handles optimization cases for Broadcast/Ramp * Broadcast/Ramp
+const Expr* mulMultilane(const Expr* lhs, const Expr* rhs, IRMutator* mutator) {
```
`ret->accept_mutator(mutator)` can be moved out of this function to places where it's called to keep the signature of this function cleaner.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1114,7 +1114,12 @@ void TensorExprKernel::lowerToBackend(BackendType backendType) {
+  try {
+    stmt = IRSimplifier::simplify(stmt);
+  } catch (const std::exception& e) {
+    std::cout << "Failed IR Simplification " << e.what() << "\n";
```
We probably should handle it somehow differently than just complaining in stdout. Just ignore if simplification failed for now? Ultimately I think simplification should never fail - it could do nothing to the stmt, but it should never throw.

nickgg replied:
There are some cases which throw due to malformed inputs, I think it's infeasible to completely ignore them, but we could move the catch into the call to simplify?

ZolotukhinM replied:
Since simplify should only be called internally, we should guarantee the well-formed inputs for it. If we have a bug somewhere, we could throw an error, but it shouldn't be a job of simplify to catch it - it should be caught one scope higher I believe (indicating that the issue is not in simplify but rather in the place that produced the malformed inputs for simplify).

ZolotukhinM reviewed (approved):
I glanced through it and it seemed reasonable to me. I left some minor comments inline and I also have one question:
Does it currently support terms consisting of non-add/mul/sub/var/imm expressions? One thing I'm particularly interested in is Div/Mod. More specifically, I don't expect simplifier to be super smart about optimizing general divs, but I do hope that we would be able so simplify, e.g.
```
2*x/N*N - x/N*N ==> x/N*N
(x%N) - (x%N - 1) ==> 1
```

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_simplifier.h:
```diff
@@ -9,17 +11,260 @@ namespace torch {
 namespace jit {
 namespace tensorexpr {
 
```
One other thing: there is a lot of code here and it is not-trivial. It might be worth adding a high-level comment here to explain what it does and roughly how it does what it does.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_printer.cpp:
```diff
@@ -368,9 +370,34 @@ void IRPrinter::visit(const Cond* v) {
+    }
+    first = false;
+    t->accept(this);
+    os() << "\n";
```
This probably screws up printing of other stuff. So far the invariant has been that statements print end-of-line and handle indentation and expressions are printed in one line.

nickgg replied:
Yes sorry this was left in debugging, will remove.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_printer.cpp:
```diff
@@ -198,7 +200,7 @@ template <
     std::enable_if_t<!std::is_floating_point<T>::value>* = nullptr>
 static void formatImm(std::ostream& os, T v) {
-  os << v;
+  os << +v;
```
Oh, I actually didn't know that it's the syntax for printing chars as integers. Please ignore my previous comment :)

nickgg commented:
To answer your question about simplifying Div/Mod, yes and no. We don't simplify Div and Mod yet (it requires some additional infrastructure I didn't want to add to this big PR).

But, neither example you gave requires Div/Mod simplification (they are Sub simplification over opaque terms) so those should both work.

```
2*x/N*N - x/N*N ==> Sub(Term(2, x/N, N). Term(1, x/N, N))  => Term(1, x/N, N) =>  x/N*N
(x%N) - (x%N - 1) ==> Sub(Polynomial(0, x%N), Polynomial(1, x%N)) => 1
```

I can add unit tests for those cases.

ZolotukhinM commented:
> To answer your question about simplifying Div/Mod, yes and no. We don't simplify Div and Mod yet (it requires some additional infrastructure I didn't want to add to this big PR).
> 
> But, neither example you gave requires Div/Mod simplification (they are Sub simplification over opaque terms) so those should both work.

Yeah, that's exactly what I wanted - I didn't expect us to be super smart about looking into Divs/Mods, but I didn't want us to bail out the moment we see them somewhere. As long as we can consider them opaque exprs and are able to check such exprs for equality, we should be good.

nickgg commented:
@ZolotukhinM addressed your comments. Please take a look.

Good thing you mentioned those two cases, because they both showed new bugs when I wrote the test (unrelated to Div and Mod). Any other ideas for test coverage?

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_printer.cpp:
```diff
@@ -1,5 +1,7 @@
 #include <torch/csrc/jit/tensorexpr/ir_printer.h>
 
+#include "torch/csrc/jit/tensorexpr/ir_simplifier.h"
```
This is still relevant :)

nickgg replied:
huh thought I got this

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -0,0 +1,1086 @@
+namespace jit {
+namespace tensorexpr {
+
+tensorexpr::SimplifierHashType Term::hashVars() const {
```
Nit: `tensorexpr::` is redundant.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -0,0 +1,1086 @@
+  // order of ops important for float
+  if (dtype().is_floating_point()) {
+    return;
+  }
```
I don't think we're preserving the order of floats since we're picking them from an `unordered_map` when putting them to `variables_`. Am I missing something?

nickgg replied:
Floating point Exprs (even opaque ones) are never inserted into a Term or Polynomial, so we should be fine. I'll change this to throw an exception since it should be a logic error.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_simplifier.h:
```diff
@@ -452,12 +332,40 @@ class IRSimplifier : public IRMutator {
+  const Expr* mutate(const Polynomial* v);
+};
+
+class IRSimplifier {
```
Probably need to `TORCH_API`.

ZolotukhinM reviewed (approved):
I didn't look into all the details of this PR, I trust tests that you have - and since it's not widely used yet we can fix any bugs in it later.

One suggestion for the tests to add is a float computation in which we should not change the order or even simplify anything - I suspect we might be doing it incorrectly now because of iterating over `unordered_map`.

nickgg commented:
Added some regression tests for float reordering.
